service/common/src/main/java/org/apache/polaris/service/events/listeners/FileBufferPolarisEventListener.java

```diff
@@ -0,0 +1,150 @@
+import org.apache.polaris.core.entity.PolarisBaseEntity;
+import org.apache.polaris.core.entity.PolarisEntity;
+import org.apache.polaris.core.entity.PolarisEntitySubType;
+import org.apache.polaris.core.entity.PolarisEntityType;
+import org.apache.polaris.core.entity.PolarisEvent;
+import org.apache.polaris.core.entity.PolarisTaskConstants;
+import org.apache.polaris.core.persistence.MetaStoreManagerFactory;
+import org.apache.polaris.core.persistence.PolarisMetaStoreManager;
+import org.apache.polaris.core.persistence.dao.entity.EntityResult;
+import org.apache.polaris.service.task.TaskExecutor;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.ObjectOutputStream;
+import java.util.HashMap;
+import java.util.Map;
+
+/** Event listener that buffers in memory and then dumps to persistence. */
+@ApplicationScoped
+@Identifier("persistence-file-buffer")
+public class FileBufferPolarisEventListener extends PersistencePolarisEventListener {
+    private static final Logger LOGGER = LoggerFactory.getLogger(FileBufferPolarisEventListener.class);
+    MetaStoreManagerFactory metaStoreManagerFactory;
+    PolarisConfigurationStore polarisConfigurationStore;
+    TaskExecutor taskExecutor;
+
+    // Value is a Pair of the first PolarisEvent that is written to the file
+    // (used for checking the first event's timestamp on this shard) and the file handle itself
+    private final HashMap<Integer, BufferShard> buffers = new HashMap<>();
+    private static final int bufferShardCount = 10;
+    private final int timeToFlush;
+
+
+    @Inject
+    public FileBufferPolarisEventListener(
+            MetaStoreManagerFactory metaStoreManagerFactory,
+            PolarisConfigurationStore polarisConfigurationStore,
+            TaskExecutor taskExecutor
+    ) {
+        this.metaStoreManagerFactory = metaStoreManagerFactory;
+        this.polarisConfigurationStore = polarisConfigurationStore;
+        this.taskExecutor = taskExecutor;
+        this.timeToFlush = polarisConfigurationStore.getConfiguration(null, FeatureConfiguration.EVENT_BUFFER_TIME_TO_FLUSH);
+
+        for (int i = 0; i < bufferShardCount; i++) {
+            buffers.put(i, createShard(i));
+        }
+    }
+
+    @Override
+    void addToBuffer(PolarisEvent polarisEvent, CallContext callCtx) {
+        long ingestionTime = System.currentTimeMillis();
+        int shardNum = polarisEvent.hashCode() % bufferShardCount;
+        BufferShard bufferShard = buffers.get(shardNum);
+        if (bufferShard == null) {
+            LOGGER.error("No buffer shard found for shard #{}. Event dropped: {}", shardNum, polarisEvent);
+            return;
+        }
+        if (bufferShard.polarisEvent == null) {
+            bufferShard.polarisEvent = polarisEvent;
+        }
+        try {
+            ObjectOutputStream oos = bufferShard.oos;
+            oos.writeObject(polarisEvent);
```

Comment:
This is not the most stable serde; we might want to explore other options

Reply:
Implemented using `Kryo` in the next revision.